### PR TITLE
feat(metrics): transform a "raw" Amplitude event into a HTTP payload

### LINKED ISSUE #4700: fxa-metrics-processor: migrate over common Amplitude event transforms from fxa-shared into message processor
There are Amplitude event transforms in fxa-shared.  Copy/consolidate them to fxa-metrics-processor.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXA-1447)
┆Issue Number: FXA-1447

### PR DESCRIPTION
**Pull request by chenba**
This patch introduces a function that takes a services map, events map,
and fuzzy events map, and returns a function that accepts a "raw"
Amplitude event and its context and produces the payload for an
Amplitude HTTP API request.

Fixes #4700 (FXA-1447)

### CODE REVIEW

**Comment by chenba** [comment]
#### Note to Reviewer(s)

I plan to merge this after PR #5009, which adds a couple things I need: configs and Dockerfile.  So there will be a couple of small edits after the review (if this gets reviewed before #5009 is merged).

File: packages/fxa-metrics-processor/src/lib/amplitude/transforms/amplitude-event.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  prune
+} from './common';
+import { getVersion } from './version';
+import { mapOs, mapDeviceModel } from './user-agent';
+import { parse } from '../../../../../fxa-shared/metrics/user-agent';
+import { mapAmplitudeUserProperties } from './user-properties';
+import { createAmplitudeEventPropertiesMapper } from './event-properties';
+
+// @TODO get from configurations
+// const { HMAC_KEY } = process.env;
```
Comment (chenba) [inline packages/fxa-metrics-processor/src/lib/amplitude/transforms/amplitude-event.ts:28]: Configs are added in PR #5009.

File: packages/fxa-metrics-processor/src/lib/amplitude/transforms/amplitude-event.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import {
+  sha256Hmac,
+  tee,
+  mapLocation,
+  createServiceNameAndClientIdMapper,
+  prune
+} from './common';
+import { getVersion } from './version';
+import { mapOs, mapDeviceModel } from './user-agent';
+import { parse } from '../../../../../fxa-shared/metrics/user-agent';
```
Comment (chenba) [inline packages/fxa-metrics-processor/src/lib/amplitude/transforms/amplitude-event.ts:23]: Dockerfile will be merged with #5009.  I'll edit it to copy `user-agent` in.

File: packages/fxa-metrics-processor/package.json [hunk 3]
```diff
@@ -39,8 +39,9 @@
     "typedi": "^0.8.0"
   },
   "devDependencies": {
+    "@types/chai": "^4.2.11",
     "@types/mocha": "^7.0.2",
-    "@types/node": "^13.11.1",
+    "@types/node": "^12.12.7",
```
Comment (6a68) [inline packages/fxa-metrics-processor/package.json:44]: Why the downgrade?

File: packages/fxa-metrics-processor/src/lib/amplitude/transforms/types.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
```
Comment (6a68) [inline packages/fxa-metrics-processor/src/lib/amplitude/transforms/types.ts:4]: I'm so excited to have some type checking to help us maintain these transforms

File: packages/fxa-metrics-processor/src/lib/amplitude/transforms/user-properties.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+import { Event, EventContext, Device } from './types';
+import { toSnakeCase } from './common';
+import { ParsedUserAgentProperties } from '../../../../../fxa-shared/metrics/user-agent';
+import { mapBrowser, BrowserProps } from './user-agent';
+
+const DAY = 1000 * 60 * 60 * 24;
+const WEEK = DAY * 7;
+const FOUR_WEEKS = WEEK * 4;
+type DevicePeriod = typeof DAY | typeof WEEK | typeof FOUR_WEEKS;
```
Comment (bbangert) [inline packages/fxa-metrics-processor/src/lib/amplitude/transforms/user-properties.ts:13]: All of these resolve to 'number', is that useful to split them out like this when it ends up being a number? Do we sometimes get this value as a DAY, WEEK, or FOUR_WEEKS? If not maybe just calling this a 'number' is fine.

File: packages/fxa-metrics-processor/package.json (hunk 3 above)
Comment (chenba) [reply packages/fxa-metrics-processor/package.json:44]: I want the types to match the node version we are using.

**Pull request merged**